Restructuring validations and workflows

aaronjae22 opened the pull request:
### **Restructuring markdown validations and updating sanity check:**

The `git diff` command now uses the `--diff-filter=ACM` option to detect files that are:

- **Added** (A): Newly added files.
- **Modified** (M): Files with changes.
- **Copied** (C): Optionally, it will capture copied files.


**How It Works:**

1. **Detect Changes**: The `git diff --name-only --diff-filter=ACM` command lists all markdown files in the articles folder that are either new or modified.
2. **Validation**: For each detected file, the `validate_markdown_metadata.py` script is run.
3. **Failure Handling**: If any file fails validation, the script exits with an error, stopping further execution.

Closes #77 

### New and modified articles are validated when PRs are opened for the main branch, and the results are visible as a check for the PR

It will achieves the following:

1. **Validate articles** when a **pull request** is opened or reopened.
2. **Validate only new or modified markdown files** within the articles folder, based on the changes in the PR.

1. **Triggering Events**:
    - **pull_request**: The workflow runs when a pull request is opened, reopened, or synchronized (when new commits are pushed to the PR).
    - **push**: The workflow also runs when code is pushed to the main branch.
2. **Detecting New and Modified Files**:
    - The step Find new and modified articles uses git diff to find files in the articles folder that are either **newly added** (A) or **modified** (M).

1. **Running Sanity Check**:
- The Run sanity check step runs your sanity_check.sh script only for the new or modified markdown files detected in the previous step.
- The script is executed with each markdown file as a parameter `./sanity_check.sh "$file"`

If no markdown files are found, the workflow exits early without running the validation script, improving efficiency.

**PR Checks:**

- When a pull request is opened, updated, or reopened, this workflow will automatically detect any new or modified articles in the articles folder and run the validation script.
- If any validation fails, the workflow will fail, and this failure will appear as a check on the PR.
- After pushing to main, the same workflow will run, ensuring that changes directly pushed to main are also validated.

Closes #78 

### Ensure any validation issues block the deployment notification to Portmap introduced in Add workflow to notify Portmap of content updates #65

To ensure that any validation issues block the deployment notification to Portmap, we can introduce a validation step before the notify-portmap job in our GitHub workflow. If the validation step fails, it will prevent the subsequent steps (including the notification to Portmap) from executing.

1. **Separate Job for Validation (validate-articles)**:
    - This job performs the validation of your articles before triggering the notification.
    - It checks out the repository, sets up Python, installs dependencies, and runs the `sanity_check.sh` script to validate new or modified markdown files.
2. **Block Notification if Validation Fails**:
    - The notify-portmap job has the line needs: validate-articles. This ensures that the notify-portmap job only runs if the validate-articles job completes successfully.
    - If any validation fails (e.g., the `sanity_check.sh` script returns an error), the entire workflow will stop, preventing the notification from being sent to Portmap.

Closes #79 

Closes #68

aaronjae22 commented on validate_markdown_metadata.py:
```diff
@@ -1,20 +1,100 @@
+        # Open the markdown file and extract content
+        with open(file_path, 'r') as f:
+            # Split the content by '---' which indicates YAML frontmatter in markdown
+            content = f.read().split('---')
```
I completely agree with this. Just one little note:

I noticed that the `extract_yaml_and_body` from Portmap has a couple of things to store the **body** markdown. e.g. `body_content = []`

I'm not sure if I should keep it that way (I can always ignore it with something like this `frontmatter, _ = extract_yaml_and_body(content)` later on) since we need it to keep it consistent with Portmap or should I modify the method and delete those **body** functionalities?

lisad replied:
I think it makes sense to leave the method the same and just discard the body.  There's always a chance we'll want to add something to this validation code that does need the body.

aaronjae22 commented:
I added **has_yaml_header** and **extract_yaml_and_body** methods from **Portmap**.

I also added `does_field_end_with_comma` method as a helper method as well as `validate_fields` that validates the required datatypes fields in the frontmatter instead of having these functionalities inside the main method.

lisad commented on validate_markdown_metadata.py:
```diff
@@ -1,20 +1,114 @@
 
-  return True if sorted_keys == KEYS else False
+# Checks if a field is of the expected type
+def is_field_valid_type(field_name, field_value, expected_type, file_path):
```
Instead of passing "file_path" in here just to be used in the ValueError text, I try to follow the pattern of only passing in the information that's needed to do the functionality.  Then, within validate_frontmatter, the utility method's exception can be caught and the additional information of file_path added at that point.

aaronjae22 replied:
I'll be taking a look at this, but I'll be merging now as we agreed yesterday
